client/grandpa-warp-sync/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Helper for handling (i.e. answering) grandpa warp sync requests from a remote peer via the
+//! [`crate::request_responses::RequestResponsesBehaviour`].
+
+use codec::Decode;
+use sc_network::config::ProtocolId;
+use sc_network::request_responses::{IncomingRequest, ProtocolConfig};
+use sc_client_api::Backend;
+use sc_finality_grandpa::GrandpaJustification;
+use sp_runtime::traits::NumberFor;
+use futures::channel::{mpsc, oneshot};
+use futures::stream::StreamExt;
+use log::debug;
+use sp_runtime::traits::Block as BlockT;
+use std::time::Duration;
+use std::sync::Arc;
+use sc_service::{SpawnTaskHandle, config::{Configuration, Role}};
+
+/// Generates the appropriate [`ProtocolConfig`] for a given chain configuration.
+pub fn protocol_config_for_chain<TBlock: BlockT, TBackend: Backend<TBlock> + 'static>(
+	config: &Configuration,
+	spawn_handle: SpawnTaskHandle,
+	backend: Arc<TBackend>,
+) -> ProtocolConfig
+	where NumberFor<TBlock>: sc_finality_grandpa::BlockNumberOps,
+{
+	let protocol_id = config.protocol_id();
+
+	if matches!(config.role, Role::Light) {
+		// Allow outgoing requests but deny incoming requests.
+		generate_protocol_config(protocol_id.clone())
+	} else {
+		// Allow both outgoing and incoming requests.
+		let (handler, protocol_config) = GrandpaWarpSyncRequestHandler::new(
+			protocol_id.clone(),
+			backend.clone(),
+		);
+		spawn_handle.spawn("grandpa_warp_sync_request_handler", handler.run());
+		protocol_config
+	}
+}
+
+const LOG_TARGET: &str = "grandpa-warp-sync-request-handler";
+
+/// Generates a [`ProtocolConfig`] for the block request protocol, refusing incoming requests.
+pub fn generate_protocol_config(protocol_id: ProtocolId) -> ProtocolConfig {
+	ProtocolConfig {
+		name: generate_protocol_name(protocol_id).into(),
+		// todo
+		max_request_size: 1024 * 1024,
```

Comment:
```suggestion
		max_request_size: 32,
```

Matches what we did in the substrate-lite PR.